In [1]:
#Import Libaries

import pandas as pd
import numpy as np
import os
import matplotlib.pyplot as plt
import seaborn as sns
import scipy

In [2]:
path = r'C:\Users\stefa\Documents\07-Instacart Basket Analysis'

In [3]:
path

'C:\\Users\\stefa\\Documents\\07-Instacart Basket Analysis'

In [4]:
df_customers = pd.read_csv(os.path.join(path, '02 Data', 'Original Data', 'customers.csv'),  index_col = False)

In [5]:
#Confirming dataframe shape
df_customers.shape

(206209, 10)

In [6]:
#Confirming dataframe
df_customers.head()

user_id First Name    Surnam  Gender       STATE  Age date_joined  \
0    26711    Deborah  Esquivel  Female    Missouri   48    1/1/2017   
1    33890   Patricia      Hart  Female  New Mexico   36    1/1/2017   
2    65803    Kenneth    Farley    Male       Idaho   35    1/1/2017   
3   125935   Michelle     Hicks  Female        Iowa   40    1/1/2017   
4   130797        Ann   Gilmore  Female    Maryland   26    1/1/2017   

   n_dependants fam_status  income  
0             3    married  165665  
1             0     single   59285  
2             2    married   99568  
3             0     single   42049  
4             1    married   40374

In [7]:
#Changing column names
df_customers.rename(columns = {'First Name': 'first_name', 'Surnam': 'last_name', 'Gender': 'gender', 'STATE': 'state', 'Age': 'age', 'n_dependants': 'number_dependents'}, inplace = True)

In [8]:
#Confirming dataframe after column name changes
df_customers.head()

user_id first_name last_name  gender       state  age date_joined  \
0    26711    Deborah  Esquivel  Female    Missouri   48    1/1/2017   
1    33890   Patricia      Hart  Female  New Mexico   36    1/1/2017   
2    65803    Kenneth    Farley    Male       Idaho   35    1/1/2017   
3   125935   Michelle     Hicks  Female        Iowa   40    1/1/2017   
4   130797        Ann   Gilmore  Female    Maryland   26    1/1/2017   

   number_dependents fam_status  income  
0                  3    married  165665  
1                  0     single   59285  
2                  2    married   99568  
3                  0     single   42049  
4                  1    married   40374

In [9]:
#Finding column datatypes
df_customers.dtypes

user_id               int64
first_name           object
last_name            object
gender               object
state                object
age                   int64
date_joined          object
number_dependents     int64
fam_status           object
income                int64
dtype: object

User_id is a system marker for an individual. No math involving them will ever be done, so changing the column to string datatype.

In [10]:
#Changing user_id datatype
df_customers['user_id'] = df_customers['user_id'].astype('str')

In [11]:
#Confirming user_id datatype change
df_customers['user_id'].dtype

dtype('O')

Data Wrangling finished. I have not been told what catagories (columns) can be removed so I'm keeping them all in. I dont think any should be removed anyway, but if the client had wanted that data removed, I would have complied.

In [12]:
df_customers.describe()

age  number_dependents         income
count  206209.000000      206209.000000  206209.000000
mean       49.501646           1.499823   94632.852548
std        18.480962           1.118433   42473.786988
min        18.000000           0.000000   25903.000000
25%        33.000000           0.000000   59874.000000
50%        49.000000           1.000000   93547.000000
75%        66.000000           3.000000  124244.000000
max        81.000000           3.000000  593901.000000

In [13]:
#Checking for missing values
df_customers.isnull().sum()

user_id                  0
first_name           11259
last_name                0
gender                   0
state                    0
age                      0
date_joined              0
number_dependents        0
fam_status               0
income                   0
dtype: int64

first_name is missing 11259 values. Those values are not important to analysis, so they remain unchanged.

In [14]:
#Checking for duplicate values
df_dups = df_customers[df_customers.duplicated()]

In [15]:
df_dups

Empty DataFrame
Columns: [user_id, first_name, last_name, gender, state, age, date_joined, number_dependents, fam_status, income]
Index: []

No duplicate values.

In [16]:
#Checking for mixed type values
for col in df_customers.columns.tolist():
  weird = (df_customers[[col]].map(type) != df_customers[[col]].iloc[0].apply(type)).any(axis = 1)
  if len (df_customers[weird]) > 0:
    print (col)   

first_name


In [17]:
df_customers['first_name'] = df_customers['first_name'].astype('str')

In [18]:
#Checking for mixed type values
for col in df_customers.columns.tolist():
  weird = (df_customers[[col]].map(type) != df_customers[[col]].iloc[0].apply(type)).any(axis = 1)
  if len (df_customers[weird]) > 0:
    print (col)  

None of the columns are mixed types.

In [19]:
ords_prods_merge = pd.read_pickle(os.path.join(path, '02 data' , 'prepared data' , 'ords_prods_merge_4.8.pkl'))

In [20]:
#Checking ords_prods_merge datatypes
ords_prods_merge.dtypes

product_id                          int64
product_name                       object
aisle_id                            int64
department_id                       int64
prices                            float64
Unnamed: 0                          int64
order_id                            int64
user_id                             int64
order_number                        int64
orders_day_of_week                  int64
orders_time_of_day (24 hours)       int64
days_since_prior_order            float64
add_to_cart_order                   int64
reordered                           int64
Check                            category
busiest days                       object
slowest days                       object
busiest_period_of_day              object
max_order                           int64
loyalty_flag                       object
avg_spending_user                   int64
spending_flag                      object
ords_freq_median                  float64
frequency_flag                    

user_id is in both dataframes. Both have the same datatypes.

In [21]:
#Removing Unnamed: 0 column
ords_prods_merge = ords_prods_merge.drop(columns=['Unnamed: 0'])

In [22]:
#Removing Check column
ords_prods_merge = ords_prods_merge.drop(columns=['Check'])

In [23]:
#Confirming Unnamed: 0 and Check columns are removed
ords_prods_merge.dtypes

product_id                         int64
product_name                      object
aisle_id                           int64
department_id                      int64
prices                           float64
order_id                           int64
user_id                            int64
order_number                       int64
orders_day_of_week                 int64
orders_time_of_day (24 hours)      int64
days_since_prior_order           float64
add_to_cart_order                  int64
reordered                          int64
busiest days                      object
slowest days                      object
busiest_period_of_day             object
max_order                          int64
loyalty_flag                      object
avg_spending_user                  int64
spending_flag                     object
ords_freq_median                 float64
frequency_flag                    object
dtype: object

In [24]:
#Changing user_id datatype in ords_prods_merge dataframe
ords_prods_merge['user_id'] = ords_prods_merge['user_id'].astype('str')

In [25]:
ords_prods_customers_merge = ords_prods_merge.merge(df_customers, on = 'user_id', indicator = True)

In [26]:
ords_prods_customers_merge.to_pickle(os.path.join(path, '02 Data','Prepared Data', 'orders_products_customers_combined.pkl'))